core/src/main/scala/cats/Eval.scala

```diff
@@ -294,12 +301,19 @@ object Eval extends EvalInstances {
    * trampoline are not exposed. This allows a slightly more efficient
    * implementation of the .value method.
    */
-  sealed abstract class Compute[A] extends Eval[A] {
+  sealed abstract class Compute[A] extends Eval[A] { self =>
     type Start
     val start: () => Eval[Start]
     val run: Start => Eval[A]
 
-    def memoize: Eval[A] = Later(value)
+    def memoize: Eval[A] =
+      new Compute[A] {
+        type Start = self.Start
+        val start: () => Eval[Start] = self.start
+        val run: Start => Eval[A] = self.run
+        override def memoize: Eval[A] = this
+        override lazy val value: A = self.value
```

Comment:
Same, this lazy val is lost on flatMap no?
Maybe memoise only applies to an outer flatMap?

I have used it when forking into two paths so I don't evaluate something twice, but this wouldn't actually work with this change I think.